cpp/include/cudf/ast/nodes.hpp

```diff
@@ -26,6 +25,21 @@
 
 namespace cudf {
 namespace ast {
+namespace detail {
+
+// Forward declaration
+class expression_parser;
+/**
+ * @brief A generic node that can be evaluated to return a value.
+ *
+ * This class is a part of a "visitor" pattern with the `linearizer` class.
+ * Nodes inheriting from this class can accept visitors.
+ */
+struct node {
```

Comment:
Are there plans to add a virtual destructor for this in a future PR or should that be in this PR?  As it is now, applications have to track and destroy each AST node by the specific node type rather than being able to collect and destroy them in a more generic `std::unique_ptr<node>` collection.

Reply:
I saw that your PRs were merged but I only looked at the implementation and didn't notice that you'd added tests, so that's good to know.

I was planning to try and handle all the non-breaking changes first, then try and stick the breaking changes together (since there will be others as well from our previous discussions) in the hope of minimizing the number of times we'd need to make changes downstream of the libcudf AST code. Does that sound OK to you?

Reply:
Sure that's fine.  I didn't think making the existing destructor virtual would break existing application code, just make it safer for application code that doesn't realize the code won't work as they expect if they try to destroy AST objects through a `node *`.

Reply:
I guess I'm thinking of source-level compatibility instead of binary compatibility. :smile:

Reply:
Ah yeah fair point, we don't need to preserve ABI either. I guess I also figured that you'd want to rewrite the JNI once when all the relevant changes land, including non breaking ones like the destructor.